activation/poetdb.go

```diff
@@ -0,0 +1,135 @@
+	xdr "github.com/nullstyle/go-xdr/xdr3"
+	"github.com/spacemeshos/go-spacemesh/common"
+	"github.com/spacemeshos/go-spacemesh/database"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/types"
+	"github.com/spacemeshos/merkle-tree"
+	"github.com/spacemeshos/poet/hash"
+	"github.com/spacemeshos/poet/shared"
+	"github.com/spacemeshos/poet/verifier"
+	"github.com/spacemeshos/sha256-simd"
+)
+
+type PoetDb struct {
+	store database.Database
+	log   log.Log
+}
+
+func NewPoetDb(store database.Database, log log.Log) *PoetDb {
+	return &PoetDb{store: store, log: log}
+}
+
+func (db *PoetDb) ValidateAndStorePoetProof(proof types.PoetProof, poetId []byte, roundId uint64, signature []byte) error {
+	root, err := calcRoot(proof.Members)
+	if err != nil {
+		return db.logError(fmt.Errorf("failed to calculate membership root for poetId %x round %d: %v",
+			poetId, roundId, err))
+	}
+	if err := validatePoet(root, proof.MerkleProof, proof.LeafCount); err != nil {
+		return db.logWarning(fmt.Errorf("failed to validate poet proof for poetId %x round %d: %v",
+			poetId, roundId, err))
+	}
+
+	var poetProof bytes.Buffer
+	if _, err := xdr.Marshal(&poetProof, proof); err != nil {
+		return db.logError(fmt.Errorf("failed to marshal poet proof for poetId %x round %d: %v",
+			poetId, roundId, err))
+	}
+
+	// TODO(noamnelke): validate signature (or extract public key and use for salting merkle hashes)
+
+	ref := sha256.Sum256(poetProof.Bytes())
+
+	batch := db.store.NewBatch()
+	if err := batch.Put(ref[:], poetProof.Bytes()); err != nil {
+		return db.logError(fmt.Errorf("failed to store poet proof for poetId %x round %d: %v",
+			poetId, roundId, err))
+	}
+	if err := batch.Put(makeKey(poetId, roundId), ref[:]); err != nil {
+		return db.logError(fmt.Errorf("failed to store poet proof index entry for poetId %x round %d: %v",
+			poetId, roundId, err))
+	}
+	if err := batch.Write(); err != nil {
+		return db.logError(fmt.Errorf("failed to store poet proof and index for poetId %x round %d: %v",
+			poetId, roundId, err))
+	}
+
+	return nil
+}
+
+func (db *PoetDb) GetPoetProofRef(poetId []byte, roundId uint64) ([]byte, error) {
+	poetRef, err := db.store.Get(makeKey(poetId, roundId))
+	if err != nil {
+		return nil, db.logInfo(fmt.Errorf("could not fetch poet proof for poetId %x round %d: %v",
+			poetId, roundId, err))
+	}
+	return poetRef, nil
+}
+
+func (db *PoetDb) GetMembershipByPoetProofRef(poetRef []byte) (map[common.Hash]bool, error) {
+	poetProofBytes, err := db.store.Get(poetRef)
+	if err != nil {
+		return nil, db.logWarning(fmt.Errorf("could not fetch poet proof for ref %x: %v", poetRef, err))
+	}
+	var poetProof types.PoetProof
+	if _, err := xdr.Unmarshal(bytes.NewReader(poetProofBytes), &poetProof); err != nil {
+		return nil, db.logError(fmt.Errorf("failed to unmarshal poet proof for ref %x: %v", poetRef, err))
+	}
+	return membershipSliceToMap(poetProof.Members), nil
+}
+
+func (db *PoetDb) logInfo(err error) error {
```

Comment:
IMO better to do it the normal way or change the logger interface all together,
its not the standard use so its a bit confusing.
more than that its not used that often.